Add use hook API reference page (#6177)
reactjs/react.dev: src/content/reference/react/components.md

## Your own components {/*your-own-components*/}

You can also [define your own components](/learn/your-first-component) as JavaScript functions.

> --

Reviewer: ```suggestion
---
```

Author: fixed

> ## Client and Server Components {/*client-and-server-components*/}

Reviewer 2: Can you make clear to people not using RSC what (if anything) to take away from this section?

Author: added a note. What do you think?

<Wip>

> [React Server Components](/learn/start-a-new-react-project#bleeding-edge-react-frameworks) are only available in React’s canary and experimental channels. Learn more about [React's release channels here](/community/versioning-policy#all-release-channels).

Author: fixed

</Wip>

> There are two main types of components in React: Server Components and Client Components. Both types follow the same basic principles of React - they are reusable, composable, and stateful - but they are used in different scenarios and have different capabilities.

Reviewer 2: did we discuss whether to capitalize these? in 'use client' page, I didn't and I would lean towards not

Author: Dan suggested that Server Components and Client Components be capitalized but I don't have a strong opinion either way. I've changed this to be lower case for now.

Reviewer 3: I'm finding it weird to be lower case. In my head, React Server Components is essentially a proper concept. It has a capitalized acronym; RSC. Similarly, SSR is Server Side Rendering, not server side rendering. 

When we've talked about RSC in past communication, like blog posts, it's been capitalized. https://react.dev/blog/2022/03/29/react-v18#server-components-is-still-in-development

> [Server Components](https://react.dev/blog/2020/12/21/data-fetching-with-react-server-components) is an upcoming feature that allows developers to build apps that span the server and client, combining the rich interactivity of client-side apps

https://react.dev/blog/2021/12/17/react-conf-2021-recap#streaming-server-rendering-with-suspense

> In the keynote, we also shared our vision for Suspense, Server Components, new React working groups, and our long-term many-platform vision for React Native.

Reviewer 2: Haven’t caught up on all the messages here but I had been thinking “React Server Components” when talking about the feature but not when talking about the type of the component, if that makes sense.

Like if it was called Full-Stack React and had server components and client components. (Maybe that’s too confusing.)

Reviewer 3: Yep, I agree! When talking about a specific server component it is lower case, when talking about React Server Components, it is upper case. I was thinking about it the same way.

When talking about a React feature being supported in a Server Component it is uppercase (but maybe that should always just be Server Component*s*)

Reviewer 2: My inclination would be:

* Frameworks that support React Server Components let you write both server components and client components.
* Only server components can use async/await at present, but client component support is planned. [this is different [this is different from what you said @TheSavior]
* I wrote a client component today and it made me sad.

i.e. capitalize it only if and only if you could replace the usage with "The React Architecture" or similar (if that were an official term)

Author: I've changed it back to capitalize "Server Components" and "Client Components" to be consistent with the React blog:

* https://react.dev/blog/2022/06/15/react-labs-what-we-have-been-working-on-june-2022#server-components
* https://react.dev/blog/2023/03/22/react-labs-what-we-have-been-working-on-march-2023#react-server-components
* https://react.dev/blog/2022/03/29/react-v18#server-components-is-still-in-development
* https://react.dev/blog/2023/05/03/react-canaries#example-react-server-components

which has more history than the react.dev doc pages that reference lowercase client component or server component:

* https://react.dev/reference/react/use-server#caveats
* https://react.dev/reference/react/useId#why-is-useid-better-than-an-incrementing-counter
* https://react.dev/reference/react/use-client#caveats

Reviewer 2: ok, sgtm – not convinced we should capitalize these long term (eg: we don't capitalize component, element, context) but seems fine for now, thanks for investigating

### Server Components {/*server-components*/}

Server Components are a new addition to React that allow developers to leverage the power of server-side rendering while maintaining the interactivity and richness of a client-side React application. They are rendered only on the server and have zero impact on the client-side bundle size, making them an excellent choice for parts of your application that are not interactive and don't need to be updated in response to user interaction.

### Client Components {/*client-components*/}

> Client Components are the traditional React components that run on the client. They can access the full range of React features, including state, effects, and access to the DOM. They are essential for creating rich, interactive features where users expect immediate feedback.

Reviewer 2: Not sure if this is temporary but we should be working towards server-first docs, not describing server components as a "new addition" and client components as "traditional"

Author: fixed.

> Client Components are the traditional React components that run on the client. They can access the full range of React features, including state, effects, and access to the DOM. They are essential for creating rich, interactive features where users expect immediate feedback.

Reviewer 4: Should we cross-link to the use client directive?

Author: Good call! Added.

## Your own components {/*your-own-components*/}

You can also [define your own components](/learn/your-first-component) as JavaScript functions.

---

## Client and server components {/*client-and-server-components*/}

<Wip>
[React server components](/learn/start-a-new-react-project#bleeding-edge-react-frameworks) are only available in React’s canary and experimental channels. Learn more about [React's release channels here](/community/versioning-policy#all-release-channels).
</Wip>

<Note>

> React server components require the use of [a supported framework](/learn/start-a-new-react-project#bleeding-edge-react-frameworks). If you're aren't working with a framework that support React server components, all the components you interact with are treated as client components.

Reviewer 5: ```suggestion
React server components require the use of [a supported framework](/learn/start-a-new-react-project#bleeding-edge-react-frameworks). If you aren't working with a framework that supports React server components, all the components you interact with are treated as client components.
```

</Note>

> There are two main types of components in React: server components and client components. Both types follow the same basic principles of React - they are reusable, composable, and stateful - but they are used in different scenarios and have different capabilities.

Author: I've removed this section since we are planning to add this information into the "use server" and "use client" directive reference pages

> There are two main types of components in React: server components and client components. Both types follow the same basic principles of React - they are reusable, composable, and stateful - but they are used in different scenarios and have different capabilities.

Author: I've removed this section since we are planning to add this information into the "use server" and "use client" directive reference pages

### Server components {/*server-components*/}

Server components allow developers to leverage the power of server-side rendering while maintaining the interactivity and richness of a client-side React application. They are rendered only on the server and have zero impact on the client-side bundle size, making them an excellent choice for parts of your application that are not interactive and don't need to be updated in response to user interaction.

### Client components {/*client-components*/}

> Client components can access the full range of React features, including state, effects, and access to the DOM. They are essential for creating rich, interactive features where users expect immediate feedback. Client component must have the [`'use client'` directive](/reference/react/use-client#use-client) place at the top of the file they are defined in.

Reviewer 5: ```suggestion
Client components can access the full range of React features, including state, effects, and access to the DOM. They are essential for creating rich, interactive features where users expect immediate feedback. Client component must have the [`'use client'` directive](/reference/react/use-client#use-client) placed at the top of the file they are defined in.
```

> In the following sections, we will delve deeper into the characteristics, capabilities, and use cases of both server components and client Components.

Author: I've removed this section since we are planning to add this information into the "use server" and "use client" directive reference pages
